### PR TITLE
README.md: refine readme

### PR DESCRIPTION
**Pull request by ruomengh**
1. Move design section to wiki
2. Add recommended configuration in installation section
3. Other small adjustment

### CODE REVIEW

**Review by eadamsintel — COMMENTED** [review]
"provides an isolated encyryption" would be more accurate

File: README.md [hunk 1]
```diff
@@ -11,37 +11,28 @@
 
 ## 1. Introduction
 
-Confidential Computing technology like Intel TDX provides isolated encryption runtime
+Confidential Computing technology like Intel® TDX provides isolated encryption runtime
 environment to protect data-in-use based on hardware Trusted Execution Environment (TEE).
 It requires a full chain integrity measurement on the launch-time or runtime environment
-to guarantee "consistently behavior in expected way" (defined by
-[Trusted Computing](https://en.wikipedia.org/wiki/Trusted_Computing)) of confidential
+to guarantee "consistently behavior in expected way" of confidential
```
Comment (eadamsintel) [inline README.md:17]: guarantee "consistent behavior in an expected way"

File: README.md [hunk 2]
```diff
@@ -11,37 +11,28 @@
...
 
 This project is designed to provide cloud native measurement for the full measurement
 chain from TEE TCB -> Firmware TCB -> Guest OS TCB -> Cloud Native TCB as follows:
 
 ![](/docs/cc-full-meaurement-chain.png)
 
-_NOTE: Different with traditional trusted computing on non-confidential environment,
+_NOTE: Different from traditional trusted computing on non-confidential environment,
 the measurement chain is not only started with Guest's `SRTM` (Static Root Of Measurement)
 but also need include the TEE TCB, because the CC VM environment is created by TEE
```
Comment (eadamsintel) [inline README.md:27]: "but it also needs to include"

File: README.md [hunk 3]
```diff
@@ -11,37 +11,28 @@
...
 via `DRTM` (Dynamic Root of Measurement) like Intel TXT on the host._
 
 From the perspective of tenant's workload, `CCNP` will expose the [CC Trusted API](https://github.com/cc-api/cc-trusted-api)
```
Comment (eadamsintel) [inline README.md:30]: "of a tenant's workload"

File: README.md [hunk 4]
```diff
@@ -11,37 +11,28 @@
...
 
 ![](/docs/ccnp-architecture-high-level.png)
 
-This project should also be able deployed on [diverse cloud native PaaS frameworks](https://www.redhat.com/en/blog/confidential-computing-use-cases) like confidential cluster, container, `kubevirt` etc.
-An example of landing architecture on confidential cluster is as follows, please
-refer [detail deployment steps](/deployment/README.md)
-
-![](/docs/ccnp-landing-confidential-cluster.png)
-
 Finally, the full trusted chain will be measured into CC report as follows using
```
Comment (eadamsintel) [inline README.md:36]: into a CC report

File: README.md [hunk 5]
```diff
@@ -11,37 +11,28 @@
...
 TDX as example:
```
Comment (eadamsintel) [inline README.md:37]: as an example

File: README.md [hunk 6]
```diff
@@ -52,50 +43,38 @@ _NOTE:_
 - The measurement of TEE, Guest's boot, OS is per CC VM, but cluster/container measurement
```
Comment (eadamsintel) [inline README.md:43]: The measurement of a TEE, guest boot flow, or OS is per CC VM  <-- I think it is the guest boot flow you are measuring.  Is it a single measurement or multiple measurements? The way it is written it sounds like a single measurement is done across those three things.  If it is more than one measurement then it would be  "measurements"  and "is" would become "are".

File: README.md [hunk 7]
```diff
@@ -52,50 +43,38 @@ _NOTE:_
...
 might be per cluster/namespace/container for cloud native architecture.
 - Please refer structure [`TDREPORT`](https://github.com/tianocore/edk2/blob/master/MdePkg/Include/IndustryStandard/Tdx.h)
```
Comment (eadamsintel) [inline README.md:45]: Please refer to the structure

File: README.md [hunk 8]
```diff
@@ -52,50 +43,38 @@ _NOTE:_
...
+- CCNP project collects container level primitives by implementing unified APIs defined in [CC Trusted API](https://github.com/cc-api/cc-trusted-api). The project will be moved to [CC Trusted API](https://github.com/cc-api/cc-trusted-api) in the near future.
```
Comment (eadamsintel) [inline README.md:46]: The CCNP project collects container level primitives by implementing the unified APIs defined..

File: README.md [hunk 9]
```diff
@@ -52,50 +43,38 @@ _NOTE:_
...
 
-- SDK is provided to simplify the use of the service interface for development,
-it covers communication to the service and parses the results from the services.
-With such SDK, users can perform related actions with one simple API call.
+### 2.1 Configuration
 
-- A `CCNP` device plugin is provided as the dependency for services such as Quote
-Server and Measurement Server. It will help with device mount and folder injection
-within the service.
+CCNP collects primitives of confidential cloud native environments running in confidential VMs, such as Intel® TD. You can setup an Intel® TDX enlightened host and then boot TD guest on it. The feasible configurations are as below.
```
Comment (eadamsintel) [inline README.md:53]: I don't think  Intel((r) TD is a valid product name. I could not find it in the brands database.

Comment (eadamsintel) [reply README.md:53]: Also it should be   "then boot a TD guest"

File: README.md [hunk 10]
```diff
@@ -52,50 +43,38 @@ _NOTE:_
...
 
-SDK PyPI package can be found [here](https://pypi.org/project/ccnp/). Please check our [documentation](https://intel.github.io/confidential-cloud-native-primitives/) for more details.
+|  CPU | Host OS  | Host packages  | Guest OS  | Guest packages  | DCAP packages |
+|---|---|---|---|---|---|
+|  Intel® Emerald Rapids | Ubuntu 22.04| Build packages referring to [here](https://github.com/intel/tdx-tools/tree/tdx-1.5/build/ubuntu-22.04) | Ubuntu 22.04 | Build packages referring to [here](https://github.com/intel/tdx-tools/tree/tdx-1.5/build/ubuntu-22.04) | [here](https://download.01.org/intel-sgx/sgx-dcap/1.19/linux/distro/ubuntu22.04-server/)
```
Comment (eadamsintel) [inline README.md:57]: It is a bit odd to have a column for the CPU and just repeat it twice. You can't merge the columns in markdown so I can't think of a better way to do this.   Is RHEL not supported?   Also, DCAP and attestation is one of the trickiest things to implement and understand.  At the very least I would add "Attestation" to the DCAP column. Not everyone will know that DCAP means attestation.  The two links go to very different places. The top link is going to be hard for people to follow. Linking to the linux pdf install guide and also linking to the package should be adequate. Do the docker instructions on the CCNP site for PCCS and QGS also work with the MVP stack? If so I would use the same links.  I also think a very brief explanation of what PCCS does and what the QGS is used for right before this table would help a lot.   You could say something like "The Platform certificate caching service (PCCS) is used to retrieve and cache PCK certificates locally to your cluster from Intel's Platform Certificate Service. This is necessary to attest the authenticity of a TDVM before a workload is started in it.  The Quote Generate Service (QGS) runs on the host in a specialized enclave to generate and use TD quotes. For convenient setup these can run inside a Docker container. Learn more at https://download.01.org/intel-sgx/sgx-dcap/1.17/linux/docs/Intel_TDX_DCAP_Quoting_Library_API.pdf"  I would embed this link in the text. 

The MVP related instructions will go away eventually so I don't think it is worth a lot of effort to update this section. If the docker CCNP instructions work for that section then link that instead.

File: README.md [hunk 11]
```diff
@@ -52,50 +43,38 @@ _NOTE:_
...
-
-- Services are designed to hide the complexity of different TEE platforms and provides
-common interfaces and scalability for cloud-native environment to address the fetching
-the fetching of quote, measurement and event log.
+## 2. Installation
 
-- SDK is provided to simplify the use of the service interface for development,
-it covers communication to the service and parses the results from the services.
-With such SDK, users can perform related actions with one simple API call.
+### 2.1 Configuration
```
Comment (eadamsintel) [inline README.md:51]: Could each of these sections apply 100% to either the host or the confidential VM?  If so I would put  "2.1 Configuration (HOST)"

"2.2 Deploy CCNP Service (Confidential VM"

"2.3 Install SDK"   I am not sure if this would be  confidential VM   or  container  I would try to write each section so that all the instructions are run 100% on a host  or 100% on a VM. It would help the user know where they should be installing things.   I don't know if this can be cleanly done when comparing confidential cluster in a VM vs confidential node as a VM  or confidential workload as a VM.  The key thing about my comment is it gets confusing for the user to know where to run the commands.

File: README.md [hunk 12]
```diff
@@ -52,50 +43,38 @@ _NOTE:_
...
-separately.
+![](/docs/ccnp-landing-confidential-cluster.png)
 
-- Quote Server: [Installation guide](service/quote-server/README.md)
-- Measurement Server: [Installation guide](service/measurement-server/README.md)
-- Event Log Server: [Installation guide](service/eventlog-server/README.md)
+### 2.3 Install SDK
 
-For SDK, user can simply install from PyPI using command:
+CCNP SDK can be used by workload for cloud native primitives collecting. It needs to be installed within the workload container image and called whenever the primitives are required. The SDK can be installed from PyPI using command:
```
Comment (eadamsintel) [inline README.md:77]: "used by a workload"

Comment (eadamsintel) [reply README.md:77]: "The SDK can be installed from PyPi using the command:"

Comment (eadamsintel) [reply README.md:77]: For the next section I would indicate that the repository needs to be cloned into the VM if you want to install that way.  I would rephrase as  "Alternatively, the CCNP can be installed from source code with the following command. Make sure to clone the repository into your confidential VM."

File: README.md [hunk 13]
```diff
@@ -108,10 +87,11 @@ cd sdk/python3
 pip install -e .
 ```
 
+### 2.4 Install CCNP Device Plugin
 For the ccnp device plugin, user can find the installation guide under the 'Installation'
```
Comment (eadamsintel) [inline README.md:91]: Follow the CCNP device plugin [installation guide ](https://github.com/intel/confidential-cloud-native-primitives/blob/ed06fdf048d34a6e765a921dba23047f84aaa043/device-plugin/ccnp-device-plugin/README.md).

**Review by eadamsintel — COMMENTED** [review]
Here is my full review

File: README.md (hunk 10 above)
Comment (ruomengh) [reply README.md:57]: Mid-stream distro also works for CentOS Stream 9, but it will not be put here until CCNP is validated with it.
Containerized PCCS and QGS work for both rows. However, it's just an alternative for mid-stream because they may release PCCS and QGS packages for Ubuntu 23.10 in future. The configuration table will be evolved along with mid-stream status.

File: README.md (hunk 6 above)
Comment (ruomengh) [reply README.md:43]: It's a single measurement for TEE + guest boot + OS because it's about a single TD.

**Comment by ruomengh** [comment]
Thanks @eadamsintel for the detailed review comments! Most of the comments have been dealt with in the latest commit.

File: README.md (hunk 10 above)
Comment (eadamsintel) [reply README.md:60]: If the same Docker instructions work for both TDX integrations then I would just use that. This table will evolve as distros update their TDX offerings.

File: README.md [hunk 14]
```diff
@@ -8,110 +8,89 @@
 ![CI Check Golang](https://github.com/intel/confidential-cloud-native-primitives/actions/workflows/pr-golang-check.yaml/badge.svg)
 ![CI Check Container](https://github.com/intel/confidential-cloud-native-primitives/actions/workflows/pr-container-check.yaml/badge.svg)
 ![CC Foundation Image Customize](https://github.com/intel/confidential-cloud-native-primitives/actions/workflows/image-rewriter.yaml/badge.svg)
+[![OpenSSF Best Practices](https://www.bestpractices.dev/projects/8325/badge)](https://www.bestpractices.dev/projects/8325)
 
 ## 1. Introduction
 
-Confidential Computing technology like Intel TDX provides isolated encryption runtime
+Confidential Computing technology like Intel® TDX provides an isolated encryption runtime
```
Comment (eadamsintel) [inline README.md:15]: I just noticed this should be  "Confidential Computer technologies"  Another way to say it would be "A Confidential Computing technology like" ..  I prefer the first

File: README.md [hunk 15]
```diff
@@ -8,110 +8,89 @@
...
 environment to protect data-in-use based on hardware Trusted Execution Environment (TEE).
 It requires a full chain integrity measurement on the launch-time or runtime environment
-to guarantee "consistently behavior in expected way" (defined by
-[Trusted Computing](https://en.wikipedia.org/wiki/Trusted_Computing)) of confidential
+to guarantee "consistent behavior in an expected way" of confidential
```
Comment (eadamsintel) [inline README.md:18]: "of a confidential computing environment"

File: README.md [hunk 16]
```diff
@@ -8,110 +8,89 @@
...
-This project should also be able deployed on [diverse cloud native PaaS frameworks](https://www.redhat.com/en/blog/confidential-computing-use-cases) like confidential cluster, container, `kubevirt` etc.
-An example of landing architecture on confidential cluster is as follows, please
-refer [detail deployment steps](/deployment/README.md)
-
-![](/docs/ccnp-landing-confidential-cluster.png)
-
-Finally, the full trusted chain will be measured into CC report as follows using
-TDX as example:
+Finally, the full trusted chain will be measured into a CC report as follows using
+TDX as an example:
```
Comment (eadamsintel) [inline README.md:38]: "Intel TDX"

File: README.md [hunk 17]
```diff
@@ -8,110 +8,89 @@
...
 
-- SDK is provided to simplify the use of the service interface for development,
-it covers communication to the service and parses the results from the services.
-With such SDK, users can perform related actions with one simple API call.
+### 2.1 Configuration for Host and Guest for Host and Guest
 
-- A `CCNP` device plugin is provided as the dependency for services such as Quote
-Server and Measurement Server. It will help with device mount and folder injection
-within the service.
+CCNP collects primitives of confidential cloud native environments running in confidential VMs(CVM), such as Intel® TDX guest. The primitives are not only from the TEE + CVM boot process + CVM OS but also from the environments running workloads, e.g. Kubernetes cluster or Docker containers. Thus, you need to check below configuration for both hosts and guests.
```
Comment (eadamsintel) [inline README.md:54]: I would add a space after "confidential VMs" and before the acronym is shown (CVM).  Also, you only have to use the circle R after Intel the first time you define Intel(R) TDX.  Then you can refer to it as Intel TDX.

**Review by eadamsintel — COMMENTED** [review]
Everything is looking a lot better. I only had some minor additional feedback.

**Comment by ruomengh** [comment]
@eadamsintel The document is updated according to your comments, except that I keep both ways of installing PCCS and QGS in the configuration table. Could you please review it again? Thanks.

File: README.md [hunk 18]
```diff
@@ -8,110 +8,88 @@
...
 chain from TEE TCB -> Firmware TCB -> Guest OS TCB -> Cloud Native TCB as follows:
 
 ![](/docs/cc-full-meaurement-chain.png)
 
-_NOTE: Different with traditional trusted computing on non-confidential environment,
+_NOTE: Different from traditional trusted computing on non-confidential environment,
 the measurement chain is not only started with Guest's `SRTM` (Static Root Of Measurement)
-but also need include the TEE TCB, because the CC VM environment is created by TEE
-via `DRTM` (Dynamic Root of Measurement) like Intel TXT on the host._
+but it also needs to include the TEE TCB, because the CC VM environment is created by TEE
```
Comment (eadamsintel) [inline README.md:28]: This is a small grammar correction I missed before   -->  "(Static Root of Mesurement), but it also needs to include the TEE TCB because the"  The comma should be after the parenthesis and before b"ut" because it is a compound sentence. The "because" is a subordinating conjunction (had to look that terminology up) so no comma is needed.  English can be very nuanced.

File: README.md [hunk 19]
```diff
@@ -8,110 +8,88 @@
...
 
 _NOTE:_
 
 - The measurement of TEE, Guest's boot, OS is per CC VM, but cluster/container measurement
 might be per cluster/namespace/container for cloud native architecture.
-- Please refer structure [`TDREPORT`](https://github.com/tianocore/edk2/blob/master/MdePkg/Include/IndustryStandard/Tdx.h)
-
-
-## 2. Design
+- Please refer to structure [`TDREPORT`](https://github.com/tianocore/edk2/blob/master/MdePkg/Include/IndustryStandard/Tdx.h)
```
Comment (eadamsintel) [inline README.md:45]: The other two bullet points ended with a period  .   so for consistency I would add a period here too.

**Review by eadamsintel — COMMENTED** [review]
I noticed two very small grammar issues.  Otherwise this is good to go.

File: README.md (hunk 18 above)
Comment (ruomengh) [reply README.md:28]: Thanks for the English learning tips :-)

File: README.md (hunk 19 above)
Comment (ruomengh) [reply README.md:45]: Good catch.

**Comment by ruomengh** [comment]
@eadamsintel Thanks a lot for all the review comments. The file has been updated.

File: README.md (hunk 18 above)
Comment (eadamsintel) [reply README.md]: I have to look these things up myself to re-remember :)

**Comment by eadamsintel** [comment]
I think it is GTG.

**Review by dongx1x — APPROVED** [review]
